### DISCUSSION TITLE
calling a plpgsql function with a composite type array argument

### DISCUSSION THREAD

**Discussion by massivefermion** [porsager/postgres#420] [Q&A] [answered]
I know this seems like a very niche and weird thing to do, but I'm trying to do this and it's not working. For now I'm using another method, but I'm just curious whether I could make it work.
Suppose you have a table like below:
```sql
create table something (
  id        int generated by default as identity primary key, 
  username  varchar(64) not null unique, 
  age       smallint
)
```
and a type like:
```sql
create type person as (
  username  varchar(64), 
  age       smallint
)
```
and a function like:
```sql
create function create_person(people person[]) returns varchar(16) as $$
  declare person person; begin
    foreach person in array people loop
      insert into something (username, age) values (person.username, person.age);
    end loop;
    return 'done';
end $$ language plpgsql;
```
I know it seems weird, but in my case I have to check some conditions and also figure out some other variable to put in the third column for each of the rows in the same table but this is the gist of the problem.
In `psql` I can do this like:
`select create_person(array[('james',31),('jack',null)]::person[]);`
But I can't figure out how to do it in nodejs.
This is what I've tried:
``` sql`create_person(${sql(users)})` ```
but I get:

> Uncaught TypeError: str.replace is not a function

I've also tried:
``` sql`select create_person(array[${sql(users2)}]::person[])` ```
But I get the same error as above.

So I got a little creative:
``` sql`select create_person(array[${users.map(u => `('${u.username}',${u.age})`).join(',')}]::person[])` ```
But this time:

> Uncaught PostgresError: malformed record literal: "('james',16),('jack',52)"

I should say that this time I didn't have any item with age equal to null because I got the error below:

> Uncaught PostgresError: invalid input syntax for type smallint: "undefined"

So anyone has any idea how to do this? Or if this is even possible?
Thanks

**Comment by porsager** [comment, marked as answer]
Yeah, there's no good way to serialize records (yet), but you were close with your last attempt.

This one should work for you ;)

```
sql`select create_person(array[${ users.map((u, i) => sql`${ i ? sql`,` : sql``}(${ u.username }, ${ u.age })`) }]::person[])`
```

It will result in a query like this, and pass the values as parameters:
```
select create_person(array[($1, $2), ($3, $4)]::person[])
```
